src/coordination/coordinator_instance.cpp

```diff
@@ -12,72 +12,85 @@
 #ifdef MG_ENTERPRISE
 
 #include "coordination/coordinator_instance.hpp"
+#include "coordination/coordinator_exceptions.hpp"
+#include "nuraft/coordinator_state_machine.hpp"
+#include "nuraft/coordinator_state_manager.hpp"
+#include "utils/counter.hpp"
 
 namespace memgraph::coordination {
 
-CoordinatorInstance::CoordinatorInstance(CoordinatorData *data, CoordinatorClientConfig config,
-                                         HealthCheckCallback succ_cb, HealthCheckCallback fail_cb)
-    : client_(data, std::move(config), std::move(succ_cb), std::move(fail_cb)),
-      replication_role_(replication_coordination_glue::ReplicationRole::REPLICA),
-      is_alive_(true) {
-  if (!client_.DemoteToReplica()) {
-    throw CoordinatorRegisterInstanceException("Failed to demote instance {} to replica", client_.InstanceName());
+using nuraft::asio_service;
+using nuraft::cmd_result;
+using nuraft::cs_new;
+using nuraft::ptr;
+using nuraft::raft_params;
+using nuraft::srv_config;
+using raft_result = cmd_result<ptr<buffer>>;
+
+CoordinatorInstance::CoordinatorInstance()
+    : raft_server_id_(FLAGS_raft_server_id), raft_port_(FLAGS_raft_server_port), raft_address_("127.0.0.1") {
+  auto raft_endpoint = raft_address_ + ":" + std::to_string(raft_port_);
+  state_manager_ = cs_new<CoordinatorStateManager>(raft_server_id_, raft_endpoint);
+  state_machine_ = cs_new<CoordinatorStateMachine>();
+  logger_ = nullptr;
+
+  // ASIO options
+  asio_service::options asio_opts;
+  asio_opts.thread_pool_size_ = 1;  // TODO: (andi) Improve this
+
+  // RAFT parameters. Heartbeat every 100ms, election timeout between 200ms and 400ms.
+  raft_params params;
+  params.heart_beat_interval_ = 100;
+  params.election_timeout_lower_bound_ = 200;
+  params.election_timeout_upper_bound_ = 400;
+  // 5 logs are preserved before the last snapshot
+  params.reserved_log_items_ = 5;
+  // Create snapshot for every 5 log appends
+  params.snapshot_distance_ = 5;
+  params.client_req_timeout_ = 3000;
+  params.return_method_ = raft_params::blocking;
+
+  raft_server_ =
+      launcher_.init(state_machine_, state_manager_, logger_, static_cast<int>(raft_port_), asio_opts, params);
+
+  if (!raft_server_) {
+    throw RaftServerStartException("Failed to launch raft server on {}", raft_endpoint);
```

Comment:
Split the construction and startup, since startup can fail.

Reply:
But if startup fail I don't want to succeed. So this is intentional

Reply:
Prefer static factory method that returns optional 
`auto CoordinatorInstance::MakeCoordinatorInstance() -> std::optional<CoordinatorInstance>`

Reply:
More importantly a constructor should not be tied to global flags. 
Parse flags, if possible construct CoordinatorInstance, then launcher_.init. On failure std::nullopt/throw (I'd prefer nullopt).

Reply:
yep okey, will add it here https://github.com/memgraph/memgraph/pull/1693